crates/pathfinder/src/consensus/inner/p2p_task.rs

```diff
@@ -280,66 +291,133 @@ pub fn spawn(
+            );
+            Ok(None)
         }
-        ProposalPart::TransactionBatch(_) => {
+        ProposalPart::TransactionBatch(ref tx_batch) => {
             // TODO check if there is a length limit for the batch at network level
-            if parts.len() >= 2 {
-                parts.push(proposal_part);
-                // TODO send for execution
-                Ok(None)
-            } else {
-                Err(anyhow::anyhow!(
+            if parts.len() < 2 {
+                return Err(anyhow::anyhow!(
                     "Unexpected proposal TransactionBatch for height and round {} at position {}",
-                    height,
+                    height_and_round,
                     parts.len()
-                ))
+                ));
             }
+
+            let Some(validator_stage) = validator_cache.remove(&height_and_round) else {
+                return Err(anyhow::anyhow!(
+                    "No ValidatorTransactionBatchStage for height and round {}",
+                    height_and_round
+                ));
+            };
+
+            let ValidatorStage::TransactionBatch(mut validator) = validator_stage else {
+                return Err(anyhow::anyhow!(
+                    "Wrong validator stage for height and round {}",
+                    height_and_round
+                ));
+            };
+
+            let transactions = tx_batch.clone();
+            parts.push(proposal_part);
+            let validator = util::task::spawn_blocking(move |_| {
+                validator.execute_transactions(transactions)?;
+                Ok::<Box<ValidatorTransactionBatchStage>, anyhow::Error>(validator)
+            })
+            .await??;
+            validator_cache.insert(
+                height_and_round,
+                ValidatorStage::TransactionBatch(validator),
+            );
+            Ok(None)
         }
         ProposalPart::Fin(ProposalFin {
             proposal_commitment,
         }) => {
+            let Some(validator_stage) = validator_cache.remove(&height_and_round) else {
+                return Err(anyhow::anyhow!(
+                    "No ValidatorTransactionBatchStage for height and round {}",
+                    height_and_round
+                ));
+            };
+
+            let ValidatorStage::TransactionBatch(validator) = validator_stage else {
+                return Err(anyhow::anyhow!(
+                    "Wrong validator stage for height and round {}",
+                    height_and_round
+                ));
+            };
+
             parts.push(proposal_part);
             let ProposalPart::Init(ProposalInit { proposer, .. }) =
                 parts.first().expect("Proposal Init")
             else {
                 unreachable!("Proposal Init is inserted first");
             };
 
+            util::task::spawn_blocking(move |_| validator.consensus_finalize(proposal_commitment))
+                .await??;
```

Comment:
The state diff commitment for block 10 contains genesis block hash in system contract 0x1. I don't think it makes sense to verify the commitment yet.

```suggestion
            // TODO uncomment when decided upon proposal is committed to the DB
            // util::task::spawn_blocking(move |_| validator.consensus_finalize(proposal_commitment))
            //    .await??;
```